ngraph/core/reference/src/runtime/reference/fft.cpp

```diff
@@ -0,0 +1,635 @@
+                                 FFTKind fft_kind)
+                {
+                    gather_to_buffer(data, length, fft_offset, stride, buffer);
+                    if (blob_is_zero(buffer, length))
+                    {
+                        return;
+                    }
+
+                    for (int64_t k = 0; k < length; ++k)
+                    {
+                        complex_type value = complex_type(0.0f, 0.0f);
+                        for (int64_t n = 0; n < length; ++n)
+                        {
+                            value += buffer[n] * twiddle(n * k, length, fft_kind);
+                        }
+                        if (fft_kind == FFTKind::Inverse)
+                        {
+                            value /= complex_type(length, 0.0f);
+                        }
+                        data[fft_offset + k * stride] = value;
+                    }
+                }
+
+                void fft1d(int64_t length,
+                           int64_t fft_offset,
+                           int64_t stride,
+                           complex_type* data,
+                           complex_type* buffer,
+                           FFTKind fft_kind)
+                {
+                    if (is_power_of_two(length))
+                    {
+                        optimized_fft1d(length, fft_offset, stride, data, buffer, fft_kind);
+                    }
+                    else
+                    {
+                        naive_fft1d(length, fft_offset, stride, data, buffer, fft_kind);
+                    }
+                }
+
+                struct InfoForFFTCalculation
+                {
+                    std::vector<int64_t> fft_axes;
+                    std::vector<int64_t> fft_lengths;
+                    std::vector<int64_t> fft_strides;
+                    std::vector<int64_t> outer_strides;
+                    std::vector<int64_t> output_fft_strides;
+                    std::vector<int64_t> output_outer_strides;
+                    std::vector<int64_t> input_fft_lengths;
+                    std::vector<int64_t> input_fft_strides;
+                    std::vector<int64_t> input_outer_strides;
+                    int64_t fft_rank;
+                    int64_t fft_size;
+                    int64_t outer_size;
+                    int64_t buffer_size;
+                };
+
+                // This function builds information needed to calculate FFT.
+                InfoForFFTCalculation get_info_for_calculation(const Shape& input_data_shape,
+                                                               const int64_t* axes_data,
+                                                               const Shape& axes_data_shape,
+                                                               const Shape& output_shape)
+                {
+                    InfoForFFTCalculation result;
+
+                    const int64_t complex_data_rank =
```

Comment:
Why do you need to create so much temporary variables? Looks like at least for half of values temporary buffers are not required.

Reply:
These temporary variables are used to store reusable intermediate results.